compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -373,7 +373,7 @@ object Types extends TypeUtils {
     final def isNotNull(using Context): Boolean = this match {
       case tp: ConstantType => tp.value.value != null
       case tp: FlexibleType => false
-      case tp: ClassInfo => !tp.cls.isNullableClass && tp.cls != defn.NothingClass
+      case tp: ClassInfo => !tp.cls.isNullableClass && !tp.isNothingType
       case tp: AppliedType => tp.superType.isNotNull
       case tp: TypeBounds => tp.lo.isNotNull
```

Comment:
After thinking about `admitsNull`, I'm now wondering: should we check the `hi` of a type bound? For example, `T >: String <: String | Null`, I don't think `T` is "NotNull".

Reply:
The test we want is `NullType <:< tp`, but `<:<` has side effects.

Reply:
Oh, sorry, never mind, I thought this was about `admitsNull`.

For `isNotNull`, you may be right, but it didn't change (the case for `TypeBounds`). It was like that before this PR.

Reply:
Oh, but you said `admitsNull`.

For `admitsNull`, the test is `NullType <:< tp`, so `lo` is right when `tp` is a `TypeBounds`.

Reply:
Sorry, I mean the `case tp: TypeBounds` in `isNotNull`. I feel it is not correct.

Reply:
You're right. It should be `hi`, not `lo`. I've pushed a fix.